File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/function/AbstractPriceReduceFunction.java
package cz.fb.manaus.reactor.profit.progress.function;

import cz.fb.manaus.core.model.MarketPrices;
import cz.fb.manaus.core.model.SettledBet;
import cz.fb.manaus.core.model.Side;

import java.util.List;
import java.util.OptionalDouble;
import java.util.function.DoubleBinaryOperator;

public abstract class AbstractPriceReduceFunction implements ProgressFunction {

    private final Side side;
    private final DoubleBinaryOperator operator;

    protected AbstractPriceReduceFunction(Side side, DoubleBinaryOperator operator) {
        this.side = side;
        this.operator = operator;
    }

    @Override
    public OptionalDouble function(SettledBet bet) {
        MarketPrices marketPrices = bet.getBetAction().getMarketPrices().getHomogeneous(side);
        List<OptionalDouble> bestPrices = marketPrices.getBestPrices(side);
        if (bestPrices.stream().allMatch(OptionalDouble::isPresent)) {
            return bestPrices.stream()
                    .filter(OptionalDouble::isPresent)
                    .mapToDouble(OptionalDouble::getAsDouble)
                    .reduce(operator);
        } else {
            return OptionalDouble.empty();
        }
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/function/AbstractPriceReduceFunction.kt
package cz.fb.manaus.reactor.profit.progress.function

import cz.fb.manaus.core.model.RealizedBet
import cz.fb.manaus.core.model.Side

abstract class AbstractPriceReduceFunction(
        private val side: Side,
        private val operator: (Double, Double) -> Double) : ProgressFunction {

    override fun invoke(bet: RealizedBet): Double? {
        val bestPrices = bet.betAction.runnerPrices.map { it.getHomogeneous(side).bestPrice }
        return if (bestPrices.all { it != null }) {
            bestPrices.filterNotNull().map { it.price }.reduce(operator)
        } else {
            null
        }
    }
}
